url/src/lib.rs

```diff
@@ -1552,12 +1552,12 @@ impl Url {
                     parser::Input::trim_tab_and_newlines(input, vfn),
                 )
             });
+            self.restore_already_parsed_fragment(fragment);
         } else {
             self.query_start = None;
+            self.restore_already_parsed_fragment(fragment);
             self.strip_trailing_spaces_from_opaque_path();
```

Comment:
Stripping spaces should only happen when the fragment is null, so the fragment should be restored first.